IdentityCore/src/util/mac/MSIDXpcProviderCache.m

```diff
@@ -0,0 +1,256 @@
+        [self.userDefaults setInteger:cachedXpcProvider forKey:MSID_XPC_PROVIDER_TYPE_KEY];
+        self.xpcConfiguration = [[MSIDXpcConfiguration alloc] initWithXpcProviderType:cachedXpcProvider];
+    });
+}
+
+- (BOOL)isXpcProviderInstalledOnDevice
+{
+    dispatch_barrier_sync(self.synchronizationQueue, ^{
+        self.isMacBrokerXpcProviderInstalled = [self isXpcProviderExist:macBrokerAppXpcInstance];
+        self.isCompanyPortalXpcProviderInstalled = [self isXpcProviderExist:companyPortalXpcInstance];
+    });
+    
+    return self.isMacBrokerXpcProviderInstalled || self.isCompanyPortalXpcProviderInstalled;
+}
+
+- (BOOL)validateCacheXpcProvider
+{
+    if (!self.xpcConfiguration)
+    {
+        // if there is no xpcConfiguration here
+        // we will use manual logic and try best to use Xpc component from MacBrokerApp first
+        // If fails, then try to use from CompanyPortalApp
+        
+        if (self.isMacBrokerXpcProviderInstalled)
+        {
+            self.cachedXpcProviderType = MSIDMacBrokerSsoProvider;
+            return YES;
+        }
+        else if (self.isCompanyPortalXpcProviderInstalled)
+        {
+            self.cachedXpcProviderType = MSIDCompanyPortalSsoProvider;
+            return YES;
+        }
+        
+        return NO;
+    }
+    
+    // Otherwise when xpcConfiguration mismatched with the existence of the Xpc provider (CP or MacBro App), fallback to the existed one on device
+    if (![self isXpcProviderExist:self.xpcConfiguration.xpcBrokerInstanceServiceBundleId])
+    {
+        // When cached Xpc configuration is available, but out of date and the Xpc component has been removed from device
+        // Let's switch to the other Xpc configuration.
+        if (self.isMacBrokerXpcProviderInstalled && self.xpcConfiguration.xpcProviderType == MSIDCompanyPortalSsoProvider)
+        {
+            // When BrokerApp Xpc is installed and cannot find CompanyPortal Xpc, we will switch to use BrokerApp Xpc
+            self.cachedXpcProviderType = MSIDMacBrokerSsoProvider;
+        }
+        else if (self.isCompanyPortalXpcProviderInstalled && self.xpcConfiguration.xpcProviderType == MSIDMacBrokerSsoProvider)
+        {
+            // When CompanyPortal Xpc is installed and cannot find BrokerApp Xpc, we will switch to use CompanyPortal Xpc
+            self.cachedXpcProviderType = MSIDCompanyPortalSsoProvider;
+        }
+        else
+        {
+            // No backup Xpc provider available, return NO
+            return NO;
+        }
+    }
+    
+    return YES;
+}
+
+- (BOOL)isXpcProviderExist:(NSString *)xpcIdentifier
+{
+    if (@available(macOS 12.0, *))
+    {
+        NSArray <NSURL *> *appURLs = [[NSWorkspace sharedWorkspace] URLsForApplicationsWithBundleIdentifier:xpcIdentifier];
+        for (NSURL *appURL in appURLs)
+        {
+            if ([appURL.absoluteString containsString:@"Applications"] && [[NSFileManager defaultManager] fileExistsAtPath:[appURL path]])
```

Comment:
When testing XPC with the ADBMacOSTestApp, this means that we will always need to move the test app from the Derived Data build folder to the Applications folder right? Is there a way to get around this for testing? Maybe allow from outside the Applications folder when in Debug mode?